pkg/ini.go

```diff
@@ -0,0 +1,228 @@
+}
+
+
+
+
+func (ini *IniFile) LoadFromFile(path string) error {
+	data, err := os.ReadFile(path)
+
+	if err != nil {
+		return err
+	}
+
+	// after reading data file we pass it to LoadFromString
+	// to parse the content of the file directly
+	return ini.LoadFromString(string(data))
+}
+
+
+
+/* Rules of ini file:
+
+1- assume there're no global keys, every keys need to be part of a section
+2- assume the key value separator is just '='
+3- keys and values should have spaces trimmed
+4- comments are only valid at the beginning of the line
+
+*/
+
+func (ini *IniFile) LoadFromString(data string) error {
+
+	lines := strings.Split(data, "\n")
+
+	// to store last section found
+	var lastSection string
+
+	// to store key and value after spliting on '='
+	var keyValue []string
+
+	var key, value string
+
+	for _, line := range lines {
+
+		//removes any spaces, newlines, and taps from beginning and ending only
+		iniLine := strings.Trim(line, " \n\t")
+
+		// checkes for emptyline or comments
+		if len(iniLine) == 0 || iniLine[0] == ';' {
+			continue
+		}
+
+		// find a valid section
+		if iniLine[0] == '[' && iniLine[len(iniLine)-1] == ']' {
+
+			// removed [] from section
+			section := strings.Trim(iniLine, "[ ]")
+
+			// if the section ended up empty then return error and reset the structure
+			if section == "" {
+				ini.IniMap = make(map[string]map[string]string)
+				return ErrInStructure
+			}
+
+			// section valid
+			ini.IniMap[section] = make(map[string]string, 0)
+
+			// sortes last section for rest its key and values
+			lastSection = section
+			continue
+		}
+
+		// if we got a key for not exist section 
+		if _, ok := ini.IniMap[lastSection]; !ok {
+
+			// reset the structure and return an error
+			ini.IniMap = map[string]map[string]string{}
+			return ErrInStructure
+		}
+
+		// valid key, seperate key and value
+		keyValue = strings.Split(iniLine, "=")
+
+		// remove any spaces, newlines, and taps
+		key = strings.Trim(keyValue[0], " \n\t")
+		value = strings.Trim(keyValue[1], " \n\t")
+
+		ini.IniMap[lastSection][key] = value
+	}
+	return nil
+}
+
+// iterate over ini structure and retrive all sections that found
+func (ini *IniFile) GetSectionNames() []string {
+	var ans []string
+
+	for section := range ini.IniMap {
+		ans = append(ans, section)
+	}
+
+	return ans
+}
+
+// retrive ini structure with all its sections, keys and values
+func (ini *IniFile) GetSections() map[string]map[string]string {
+	return ini.IniMap
+}
+
+
+
+func (ini *IniFile) Get(section_name, key string) (string, error) {
+
+	// check if given section_name is exist in ini structure and return error if not
+	if _, ok := ini.IniMap[section_name]; !ok {
+		return "", ErrNoSection
+	}
+
+	// check if given key is exist in ini structure and return error if not
+	if _, ok := ini.IniMap[section_name][key]; !ok {
+		return "", ErrNoKey
+	}
+
+	// return the value of given section_name and key
+	return ini.IniMap[section_name][key], nil
```

Comment:
It should return a boolean instead of error to indicate whether it exists or not
```suggestion
    value, exists := ini.IniMap[section][key]
    return value, exists
```